src/bridge/cockpitdisksamples.c

```diff
@@ -153,3 +158,130 @@ cockpit_disk_samples (CockpitSamples *samples)
   g_strfreev (lines);
   g_free (contents);
 }
+
+static gchar*
+read_file (const gchar *path)
+{
+  g_autoptr(GError) error = NULL;
+  gchar *contents = NULL;
+  if (!g_file_get_contents (path, &contents, NULL, &error))
+    {
+      /* Do not log:
+       * - file does not exist: process died and can no longer be sampled
+       * - access denied: bridge is not privileged and can't sample all processes
+       * - sometimes ERROR_FAILED is returned with message: "No such process"
+       */
+      if (g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT) ||
+          g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_ACCES) ||
+          (g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_FAILED) && g_strrstr(error->message, "No such process")))
+        {
+          return NULL;
+        }
+
+      g_warning ("Domain: %s, code: %d, message: %s", g_quark_to_string (error->domain), error->code, error->message);
+      return NULL;
+    }
+
+  return contents;
+}
+
+static void
+table_add_values(GHashTable *table,
+                   gchar *cgroup,
+                   guint64 read,
+                   guint64 write)
+{
+  cgroup_values_t *values = g_hash_table_lookup (table, cgroup);
+  if (values == NULL)
+    {
+      values = g_new0(cgroup_values_t, 1);
+      g_hash_table_insert (table, g_strdup (cgroup), values);
+    }
+
+  values->read = read;
+  values->write = write;
+}
+
+static void
+get_process_io (const gchar *dir_name,
+                GHashTable *table)
+{
+  g_autofree gchar *path = g_strdup_printf ("/proc/%s/io", dir_name);
```

Comment:
The extremely-theoretical race between reading `/io` and `/cgroup` tells me that you ought to be using dirfds, but I think we could let it slide...